presto-main/src/main/java/com/facebook/presto/cost/AggregationStatsRule.java

```diff
@@ -76,27 +94,59 @@ public static PlanNodeStatsEstimate groupBy(PlanNodeStatsEstimate sourceStats, C
             }));
         }
 
-        double rowsCount = 1;
+        double rowsCount = getDefaultAggregateSelectivityCoefficient(session);
+        double totalSize = 0;
         for (VariableReferenceExpression groupByVariable : groupByVariables) {
             VariableStatsEstimate symbolStatistics = sourceStats.getVariableStatistics(groupByVariable);
             int nullRow = (symbolStatistics.getNullsFraction() == 0.0) ? 0 : 1;
             rowsCount *= symbolStatistics.getDistinctValuesCount() + nullRow;
+            if (!isNaN(symbolStatistics.getAverageRowSize())) {
+                totalSize += symbolStatistics.getAverageRowSize() * rowsCount;
+            }
+        }
+        // Set estimation boundary for aggregation:
+        // If the source only return a single row, it's guaranteed the upper bound is a single row output for this agg.
+        // Aggregation node cannot return output row count more than estimated row count of source node.
+        // So if source stats is present, use the output row count from the source stats as the estimated output row
+        // count as the upper bound limit to avoid returning an unknown output potentially breaks the join search space
+        double outputRowCountFromSource = getDefaultAggregateSelectivityCoefficient(session) * sourceStats.getOutputRowCount();
+        if (!isNaN(rowsCount)) {
+            result.setOutputRowCount(min(rowsCount, outputRowCountFromSource));
```

Comment:
I think we want to bound by the total size of the source, not the source * the selectivity coefficient.

Reply:
initial thinking was to take the selectivity factor in across all stats even for the output estimation from source because if the selectivity factor 90%, it'll be a discounted upper bound limit for this aggregate operator, is it the right assumption? If not I can change the code back to ignore taking the selectivity factor into account.